### ISSUE TITLE
Manual device placement for distributed training isn't async like TF1

### ISSUE THREAD

**Issue by Apprisco** [tensorflow/tensorflow#56880]
### Issue Type

Bug

### Source

binary

### Tensorflow Version

2.9.12

### Custom Code

No

### OS Platform and Distribution

Windows 11

### Python version

3.9.12

### CUDA/cuDNN version

11.2/8.6

### GPU model and memory

3090, 24GB

### Current Behaviour?

```shell
As I am not able to use a tensorflow dataset for my specific application (migrated from v1, would rather keep the custom loader that works in parallel very well, tuned to specific machines.), I was looking into V1-esque distributed training strategies.
In V1, we would simply loop over gpus and redo the computation while splitting the inputs in quarters.
That is what I did in tf2, and that is what the manual placement guide says.
.debugging.set_log_device_placement(True)

gpus = tf.config.list_logical_devices('GPU')
if gpus:
  # Replicate your computation on multiple GPUs
  c = []
  for gpu in gpus:
    with tf.device(gpu.name):
      a = tf.constant([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]])
      b = tf.constant([[1.0, 2.0], [3.0, 4.0], [5.0, 6.0]])
      c.append(tf.matmul(a, b))

  with tf.device('/CPU:0'):
    matmul_sum = tf.add_n(c)

  print(matmul_sum)
However, the amount of time the computation takes on multiple gpus is simply longer than the time that takes on a single gpu.
So I printed what happens after each gpu's computations are finished, and as I expected, the processes are not running async.
I even accumulated the gradients to compile them later, but it did not fix the problem.

Is there any way to make distributed training work without a tf.dataset? I've already gone through so many issues to make this work, including the following (not even a distributed issue)
https://github.com/tensorflow/tensorflow/issues/56878
```

### Standalone code to reproduce the issue

```shell
.debugging.set_log_device_placement(True)

gpus = tf.config.list_logical_devices('GPU')
if gpus:
  # Replicate your computation on multiple GPUs
  c = []
  for gpu in gpus:
    with tf.device(gpu.name):
      a = tf.constant([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]])
      b = tf.constant([[1.0, 2.0], [3.0, 4.0], [5.0, 6.0]])
      c.append(tf.matmul(a, b))

  with tf.device('/CPU:0'):
    matmul_sum = tf.add_n(c)

  print(matmul_sum)
```

**Comment by mohantym** [comment]
Hi @Apprisco !
Have you referred [distribution-training ](https://www.tensorflow.org/guide/distributed_training)guide yet? You can declare desired GPU devices in strategy and then run the functions under strategy.scope()
```
mirrored_strategy = tf.distribute.MirroredStrategy(devices=["/gpu:0", "/gpu:1"])
with mirrored_strategy.scope():
   // Your operations here

```

**Comment by Apprisco** [comment]
I have. Again, as I mentioned, distributed training strategies aren't an option: we've already tried eons to get them to work.
1. Cannot split dataset via experimental split as our dataset is loaded via batches of numpy arrays, experimental dataset requires the whole dataset to be made with tf.Dataset.
2. to make the mirrored strategy not go into crazy tf.function retracing mode, we need to wrap the strat.run() in a tf.function.
Problem with this is the issue I mentioned: https://github.com/tensorflow/tensorflow/issues/56878
Basically, mirrored variables don't seem to generate gradients, which is still an open issue and I haven't found any solution.
Those mirrored variables will simply not give any gradient output that isn't none, and thus w/o using gradienttape it doesn't take up any vram, while with using gradient tape it spits out errors. Weird thing is that this happens even outside of mirrored strategy scope, so maybe it's just something to do with tf.function itself.
3. I want the tf1 version, aka manual gpu computation replication to work as intended. If it's not a working feature, why is it even mentioned in the tutorial? It is also outrageous that distribute strategies are the only built-in functionality for multi-gpu training and now requires tf.datasets. Horovud seems like a much better option now than any sort of distributed strategy from Tensorflow.

**Comment by mohantym** [comment]
Ok @Apprisco !
Thanks for the update. @gadagashwini ! Could you please look at this issue.

**Comment by gadagashwini** [comment]
Hi @Apprisco, Seems i don’t observe any time constraint on executing above code on multiple GPUs on Ubuntu.
Take a look at below screenshot. Please confirm the  
<img width="1565" alt="Screen Shot 2022-07-25 at 12 52 45 PM" src="https://user-images.githubusercontent.com/99852755/180721149-f5ceb2a6-50f3-4d60-9bc0-28170aaf83af.png">

Please confirm if you still see issue persists. Thank you!

**Comment by Apprisco** [comment]
The problem is that it's not running async, a more complex computation shows it easily. You can also just put gpu print statements to see which computation finishes first, etc.
This is still definitely an issue: here's the code that specifically doesn't work for me.
```
accumulate_disc_grad=[]
accumulate_gen_grad=[]
for gpu in gpus:
    with tf.device(gpu.name):
        disc_grad=get_disc_gradients(x,y)
        gen_grad=get_gen_gradients(x)
        # accumulate gradients however
disc_opt..apply_gradients(zip(accumulate_disc_grad,self.discc.trainable_variables))
```
This is in no way async: after gpu 0 has finished computation, it moves on to gpu 1.
Then gpu 2, then gpu 3. the utilization graph shows all the story, but sadly it's hard to get an actual utilization graph off the cluster as it only logs every minute.

**Comment by Apprisco** [comment]
Issue has been resolved by utilizing horovod along with mpirun and NCCL, although this is still an issue.
Horovod seems to be the much better solution compared to mirroredstrategy and manual re-compute in the first place, there was no need to make any changes other than implementing HVD to existing code.
MPIrun was able to distribute the data to all 10 gpus on 5 separate nodes without issues.

**Comment by gadagashwini** [comment]
Hi @Apprisco, Can we close if the issue is resolved? Thank you!
